File: Back-End/backend/src/models/Watchlist.ts
import * as admin from 'firebase-admin';
import { db } from '../global/firebase';

// ── 型別定義 ────────────────────────────────────────────────────────────────

export interface WatchlistInput {
  stockId: string;
  targetPrice: number;
  note?: string;
}

export interface WatchlistDoc extends Required<WatchlistInput> {
  createdAt: Date;
  updatedAt: Date;
}

export interface WatchlistPatch {
  targetPrice?: number;
  note?: string;
}

const COL = 'watchlist';

// ── Model ───────────────────────────────────────────────────────────────────

export class Watchlist {
  static async findAll(): Promise<WatchlistDoc[]> {
    const snap = await db.collection(COL).orderBy('created_at', 'asc').get();
    return snap.docs.map(d => deserialize(d));
  }

  static async findByStockId(stockId: string): Promise<WatchlistDoc | null> {
    const doc = await db.collection(COL).doc(stockId).get();
    return doc.exists ? deserialize(doc) : null;
  }

  /** 新增關注股票（Document ID = stock_id，已存在時回傳 null） */
  static async create(input: WatchlistInput): Promise<WatchlistDoc | null> {
    const ref      = db.collection(COL).doc(input.stockId);
    const existing = await ref.get();
    if (existing.exists) return null;

    await ref.set({
      stock_id:     input.stockId,
      target_price: input.targetPrice,
      note:         input.note ?? '',
      created_at:   admin.firestore.FieldValue.serverTimestamp(),
      updated_at:   admin.firestore.FieldValue.serverTimestamp(),
    });
    const created = await ref.get();
    return deserialize(created);
  }

  /** 更新目標價 / 備註 */
  static async update(stockId: string, patch: WatchlistPatch): Promise<WatchlistDoc | null> {
    const ref      = db.collection(COL).doc(stockId);
    const existing = await ref.get();
    if (!existing.exists) return null;

    const data: Record<string, unknown> = {
      updated_at: admin.firestore.FieldValue.serverTimestamp(),
    };
    if (patch.targetPrice !== undefined) data.target_price = patch.targetPrice;
    if (patch.note        !== undefined) data.note         = patch.note;

    await ref.update(data);
    const updated = await ref.get();
    return deserialize(updated);
  }

  static async delete(stockId: string): Promise<boolean> {
    const ref = db.collection(COL).doc(stockId);
    const doc = await ref.get();
    if (!doc.exists) return false;
    await ref.delete();
    return true;
  }
}

// ── 反序列化 ─────────────────────────────────────────────────────────────────

function deserialize(doc: admin.firestore.DocumentSnapshot): WatchlistDoc {
  const d = doc.data()!;
  return {
    stockId:     doc.id,
    targetPrice: d.target_price,
    note:        d.note ?? '',
    createdAt:   d.created_at instanceof admin.firestore.Timestamp
      ? d.created_at.toDate()
      : new Date(),
    updatedAt:   d.updated_at instanceof admin.firestore.Timestamp
      ? d.updated_at.toDate()
      : new Date(),
  };
}

File: Back-End/shioaji-service/models/watchlist.py
from __future__ import annotations
import asyncio
from typing import Optional
from google.cloud.firestore import SERVER_TIMESTAMP
from lib.firebase import get_db

COL = "watchlist"


def _ts_to_iso(ts) -> Optional[str]:
    return ts.isoformat() if hasattr(ts, "isoformat") else None


def _from_doc(doc) -> dict:
    d = doc.to_dict()
    return {
        "stockId": doc.id,
        "targetPrice": d.get("target_price", 0),
        "note": d.get("note", ""),
        "createdAt": _ts_to_iso(d.get("created_at")),
        "updatedAt": _ts_to_iso(d.get("updated_at")),
        "sortIndex": d.get("sort_index", 0),
    }


async def find_all() -> list[dict]:
    def _run():
        snap = get_db().collection(COL).get()
        docs = [_from_doc(d) for d in snap]
        return sorted(docs, key=lambda w: w["sortIndex"])
    return await asyncio.to_thread(_run)


async def find_by_stock_id(stock_id: str) -> Optional[dict]:
    def _run():
        doc = get_db().collection(COL).document(stock_id).get()
        return _from_doc(doc) if doc.exists else None
    return await asyncio.to_thread(_run)


async def create(stock_id: str, target_price: float, note: str = "") -> Optional[dict]:
    def _run():
        ref = get_db().collection(COL).document(stock_id)
        if ref.get().exists:
            return None
        ref.set({
            "stock_id": stock_id,
            "target_price": target_price,
            "note": note,
            "created_at": SERVER_TIMESTAMP,
            "updated_at": SERVER_TIMESTAMP,
        })
        return _from_doc(ref.get())
    return await asyncio.to_thread(_run)


async def update(stock_id: str, patch: dict) -> Optional[dict]:
    def _run():
        ref = get_db().collection(COL).document(stock_id)
        if not ref.get().exists:
            return None
        data: dict = {"updated_at": SERVER_TIMESTAMP}
        if "targetPrice" in patch: data["target_price"] = patch["targetPrice"]
        if "note" in patch:        data["note"]         = patch["note"]
        ref.update(data)
        return _from_doc(ref.get())
    return await asyncio.to_thread(_run)


async def delete(stock_id: str) -> bool:
    def _run():
        ref = get_db().collection(COL).document(stock_id)
        if not ref.get().exists:
            return False
        ref.delete()
        return True
    return await asyncio.to_thread(_run)


async def reorder(order: list[str]) -> None:
    def _run():
        db = get_db()
        batch = db.batch()
        for index, stock_id in enumerate(order):
            batch.update(db.collection(COL).document(stock_id), {"sort_index": index})
        batch.commit()
    await asyncio.to_thread(_run)
